[Sonoff Pow] Add support for HLW8012 by @papperone

TD-er opened the pull request:
Testing support, just for use with `#define PLUGIN_SET_SONOFF_POW`

Since it is a test setup, also a very good use case to test Arduino ESP8266 core 2.4.0 as default

Initial commit + PR #690 was by @papperone

TD-er committed:
[Nextion] Initial support for Nextion displays

Code taken from playground P117 and added minimal changes.
Still need some work to check and test with real display.
Not tested yet, myself.

TD-er committed:
Merge remote-tracking branch 'papperone/mega' into merge/papperone

Remove edits on P073 to separate PRs

TD-er committed:
Merge remote-tracking branch 'papperone/mega' into merge/papperone

Remove edits on P073 to separate PRs

TD-er committed:
Merge remote-tracking branch 'papperone/mega' into merge/papperone

Remove edits on P073 to separate PRs

TD-er committed:
Merge remote-tracking branch 'papperone/mega' into merge/papperone

Remove edits on P073 to separate PRs

GHolli committed:
It seems the offsets [3]..[4] for the detection range setting (command byte 0x99) are wrong in the latest online data sheet: http://www.winsen-sensor.com/d/files/infrared-gas-sensor/mh-z19b-co2-ver1_0.pdf
According to the MH-Z19B data sheet version 1.2, valid from: 2017.03.22 (received 2018-03-07), the offset should be [6]..[7] instead.
Fixing the implementation accordingly.

GHolli committed:
Update _P049_MHZ19.ino

GHolli committed:
Change comma to semicolon

Same functionality, but more common.

Grovkillen committed:
Updated JSON path for sensor values and names

+file type filter on upload (not OTA since that is not part of ESP Easy core)

Grovkillen committed:
Delete WebServer.ino

Same mistake as before, wrong folder upload

Grovkillen committed:
Moved to correct folder

Grovkillen committed:
Added value number as discussed.

TD-er committed:
[ESP32] Event driven WiFi for ESP32

Not tested, just got some inspiration by these files:

- https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/examples/WiFiClientEvents/WiFiClientEvents.ino
- https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/src/WiFiGeneric.cpp
- https://github.com/espressif/arduino-esp32/blob/master/tools/sdk/include/esp32/esp_event.h

TD-er committed:
[Sonoff Pow] Add support for HLW8012

TD-er committed:
[Sonoff Pow] Add support for HLW8012

TD-er committed:
[Sonoff Pow] Add support for HLW8012

TD-er committed:
[Sonoff Pow] Add support for HLW8012

TD-er commented:
Not yet to be merged, since it is a very specific build, preparing some defaults may be nice to add.
And it is still a test setup, so not yet to be used in the releases.

```
Blue led:   PWM2 => MTDO => GPIO15
Red/relais: PWM0 => MTDI => GPIO12

Switch:      E-FW         => GPIO0

HLW8012:  CF1 PWM1   MTCK => GPIO13
          CF  SCL    MTMS => GPIO14
          SEL IO5            GPIO5
          
          
SDA:                         GPIO2
```

Also need [ESPeasy wiki page](https://www.letscontrolit.com/wiki/index.php/Sonoff_basic) for details on the configuration of the switches.
And the default pins for I2C conflict with the pins needed for the HWL8012

psy0rz committed:
automaticly updated release notes for mega-20180328

Grovkillen commented:
I wish I had the hardware so I could test and document this.

linuxnico committed:
add ds1822 definition to ds_read_temp function

martinusGH committed:
Merge pull request #1180 from TD-er/bugfix/esp32_eventbased_wifi

[ESP32] Event driven WiFi for ESP32

TD-er committed:
[Arduino IDE] Revert changes in include paths

Were changed to be able to test the PlatformIO feature to compile only files that are actually used.
Apparently the Arduino IDE cannot find the files anymore because that IDE needs the parent directory to be named as the main project file, not "src"

TD-er committed:
Merge pull request #1189 from TD-er/bugfix/include_paths_ArduinoIDE

[issue #1188] Revert changes in include paths

psy0rz committed:
automaticly updated release notes for mega-20180329

clumsy-stefan committed:
_P015_TSL2561 added Broadband and IR Values

Extended Plugin to also read Broadband and IR values from device.

GHolli committed:
Downstream updates (#1)

* Initial re-organisation of plugin sets (#1019)

* 12bit ADC values acceted up to 4095

MCP3221 is a 12bit ADC so accepted values should be up to 40195 not 1024

* automaticly updated release notes for v2.0-20180313

* - move all structs from ESPEasy.ino into header file
- add some fix for headers
- add const to rtttl notes array

* [issue #1066] Fix Webserver handle_control

There wasn't a proper start/end for the streaming webserver.
See discussion on issue #1066

* automaticly updated release notes for mega-20180314

* Custom controller GUI option

* Fixed #433

Not tested!

* #893, part 2 receiving

* Merge branch 'mega' of github.com:letscontrolit/ESPEasy into mega

# Conflicts:
#	src/ESPEasy.ino

* Rules fix

* #893 part1, send messages

* Custom controller GUI option

* #893, part 2 receiving

* Rules fix

* automaticly updated release notes for mega-20180315

* - move all structs from ESPEasy.ino into header file

- add some fix for headers
- add const to rtttl notes array

* Merge branch 'mega' of github.com:letscontrolit/ESPEasy into mega

# Conflicts:
#	src/ESPEasy.ino

* [v2.0 split globals] Base split on v2.0 version of ESPEasy.ino

*  MCP3221 allow values for calibration (#1073)

MCP3221 allow values for calibration

* [issue #1066] Fix Webserver handle_control (#1079)

[issue #1066] Fix Webserver handle_control

* Fixed multi instance problem on SHT3x #433 (#1081)

Fixed multi instance problem on SHT3x #433

* [issue #1057] AM PM display issue

X:XX PM (for hours 1...9)
X:XX AM (for hours 1...9)
XX:XX PM (for hours 10...12)
XX:XX AM (for hours 10...12)

Added for H:MM (without seconds) notations:
- `%systm_hm%`
-`%systm_hm_am%`

* [issue #993] variable in MQTT subscribe topic

As [noted](https://github.com/letscontrolit/ESPEasy/issues/993#issuecomment-373140135) by @kluszcyn

* [issue #951] Changed System#Initialized  => System#Wake

Changed System#Initialized  => System#Wake
New event name makes more sense and Initialized may be used later to indicate all needed connections, etc. are ready.

* [issue #1057] AM PM display issue

X:XX PM (for hours 1...9)
X:XX AM (for hours 1...9)
XX:XX PM (for hours 10...12)
XX:XX AM (for hours 10...12)

Added for H:MM (without seconds) notations:
- `%systm_hm%`
-`%systm_hm_am%`

* [issue #993] variable in MQTT subscribe topic

As [noted](https://github.com/letscontrolit/ESPEasy/issues/993#issuecomment-373140135) by @kluszcyn

* [issue #951] Changed System#Initialized  => System#Wake

Changed System#Initialized  => System#Wake
New event name makes more sense and Initialized may be used later to indicate all needed connections, etc. are ready.

* automaticly updated release notes for mega-20180316

* automaticly updated release notes for v2.0-20180316

* Add download button for current rule set

* - add encoder write #1086

* changed memory tracker to uint64_t for the ESP32

* Changed so that running all tasks immediately with deep sleep enabled only happens on first iteration after wake up

* Keep uint. Just change the init value.

* [WiFi] Try to minimize effect of WiFi interruptions

And send event when connected to a different accesspoint.
Currently the impact on WiFi reconnects is enorm. The ESP stalls, won't react on events, etc.

* - fix success setting

* - fix P073 success setting in PLUGIN_WRITE callback

* - initialize pointer in P073

* [Relaxed WiFi] Perform WiFi initialization in steps

Doing WiFi initialization in multiple steps, will reduce boot time a lot.
For deepsleep the Setup is waiting for wifi initialization to complete, but for normal use the wifi connection is initiated during setup and checked on regular intervals to see if it is available.
The same for WiFi reconnect, which will hopefully not affect the performance of the ESP module during reconnects.

Also added a check + event when connected to another accesspoint with different BSSID (MAC address of accesspoint). Also the last used BSSID is used to reconnect to try to connect to the same device over and over, which may be useful with WiFi repeaters.

* [Travis error] Some #include test that wasn't meant to be included

... in the commit.

* [JSON] Add uint name (#1099) and WiFi information (PR #1113)

As suggested by @Grovkillen on PR #1113.

* [issue #1007] Allow user specific data in separate file

File will be ignored by git and can be simply included using `-D USE_OWN_DEFAULTS` in the build

* #ifndef instead of #ifdef

* [issue #1007] Add other defines as suggested by @uzi18

* [Memory] parseTemplate did not check if markers were found

"Free Mem" reported this function to be using quite a lot of memory.
Maybe that was related to parsing texts without matching ']' or '#'
Just added some checks to see if they were found.

* [issue #1097] MQTT publish after deep sleep

The MQTTConnect was only called after a while. This is a problem for deepsleep.
So now the MQTT connect is done at the first loop when deep sleep is enabled.

* Add download button for current rule set (#1102)

Add download button for current rule set

* - add encoder write command #1086 (#1103)

- add encoder write command #1086

* Changed so that running all tasks immediately with deep sleep enabled… (#1110)

Changed so that running all tasks immediately with deep sleep enabled only happens on first iteration after wake up

* Merge pull request #1112 from uzi18/fix-p073

Fix p073 - #1111

* Merge pull request #1114 from TD-er/feature/wifi_details_json

[JSON] Add uint name (#1099) and WiFi information (PR #1113)

* Merge pull request #1113 from TD-er/feature/reduce_impact_wifi_disconnect

[Relaxed WiFi] Perform WiFi initialization in steps

* Merge pull request #1116 from TD-er/feature/allow_own_defaults

[issue #1007] Allow user specific data in separate file

* Merge pull request #1118 from TD-er/bugfix/mqtt_publish_after_deep_sleep

[issue #1097] MQTT publish after deep sleep

* [merge] MIssing #endif due to merging.

* automaticly updated release notes for mega-20180319

* automaticly updated release notes for v2.0-20180319

* Fix wiring description for SN75176 transceiver

* Revert "Merge pull request #1116 from TD-er/feature/allow_own_defaults"

This reverts commit 8ccbdcadf0c214d676973d0936d4e9132b81b083.

* Fix wiring description for SN75176 transceiver

* Revert "Merge pull request #1116 from TD-er/feature/allow_own_defaults"

This reverts commit 8ccbdcadf0c214d676973d0936d4e9132b81b083.

* [issue #1097] WiFi connect with deepsleep enabled trigger watchdog reset

Did not allow for enough time to do background tasks while waiting for wifi connect with DeepSleep enabled.

* [issue #1097] WiFi connect with deepsleep enabled trigger watchdog reset

Did not allow for enough time to do background tasks while waiting for wifi connect with DeepSleep enabled.

* automaticly updated release notes for mega-20180320

* automaticly updated release notes for v2.0-20180320

* [wifi] Setup with factory settings could not complete

* [wifi] Setup with factory settings could not complete

* [merge issue] Forgotten #endif in merge

* - P012 fix command interpretation

* - P023 fix command interpretation

* - P035 move strings to flash

* - P048 fetch params only when needed

* automaticly updated release notes for mega-20180321

* automaticly updated release notes for v2.0-20180321

* Serial WD crash

* [wifi] Make more robust checks and increase timeout

See #1145

* - P012 fix command interpretation

* - P023 fix command interpretation

* - P035 move strings to flash

* - P048 fetch params only when needed

* Serial WD crash

* [wifi] Make more robust checks and increase timeout

See #1145

* [Memory] parseTemplate did not check if markers were found

"Free Mem" reported this function to be using quite a lot of memory.
Maybe that was related to parsing texts without matching ']' or '#'
Just added some checks to see if they were found.

* [WiFi] Take much more time for initial WiFi connect during setup

Accesspoint is still active, so it may take a lot more time to make the initial connection

* [WiFi] Take much more time for initial WiFi connect during setup

Accesspoint is still active, so it may take a lot more time to make the initial connection

* automaticly updated release notes for mega-20180322

* ESP32 fix

* [Command] Parsing improvement for commands

Use more intelligent parsing of commands and switch statement to improve speed

* automaticly updated release notes for mega-20180323

* [WiFi] Reverted those temp String uses for ssid/pass

Was introduced due to misunderstanding of reported issues with WiFi.

* [issue #1131] Add %unixtime%

Show the current time as [Unix time](https://en.wikipedia.org/wiki/Unix_time) (numbers of seconds that have elapsed since midnight 1970-01-01, UTC)

* [issue #1097] Call MQTTclient.loop() during deepsleep

The `MQTTclient.loop()` function should be called quite often.
This does the actual MQTT handling.
In normal operations, this is called quite frequently, but it should also be called when deepsleep is active and right before deepsleep is activated again.

* [P1] Support for latest standard

* [wifi] Event based connect/reconnect

Now using the core WiFiEvents to act on:

- connect
- disconnect
- gotIP

This will make polling for a status no longer needed.
It also allows to get a better understanding in why diconnects happen, since the disconnect reason is given in the events.
This extra information is also present on the information panel on the webinterface and also in the JSON output.

The entire connect/disconnect/reconnect is now done completely asynchronous and thus takes much less resources from the ESP.
A disconnect will now also send an event: "WiFi#Disconnected"

For deepsleep this means there has to be a minimum awake time set and perhaps use rules to set the ESP into sleep again when done.
Typical connection times to an accesspoint are about 3 seconds + about 1 second for DHCP.

This also resolves the issues where services my try to connect while there is a WiFi connection, but not yet an IP address set.

I already noticed a lot of NTP servers from pool.ntp.org are non responsive the last few weeks, so that may still take some time, but that's another issue.

* automaticly updated release notes for mega-20180324

* [NTP] Speedup NTP again with async WiFi

Previously it would only start calling "now" when the WiFi was already setup.
And in calling `now()`, a call to an NTP server is being made. When this fails, retry again in a minute.
But now the WiFi initialization is asynchronous, so the first NTP query was already failing and thus the retry is already set to a minute.

This is fixed now and also the pool.ntp.org will pick a random one from 0...3 (.pool.ntp.org).
The DNS queries from this pool change about every minute, so if you want to retry because one of the servers is offline, you should try one from the other `N.pool.ntp.org` or else you have to wait for another minute.

* - fix for #1170 - uservar not setted on encwrite

* automaticly updated release notes for mega-20180326

* [issue #1174] Show degree sign via external writes to LCD

Fix suggested by [elmicha](https://github.com/letscontrolit/ESPEasy/issues/1174#issuecomment-376244766)

* [ESP32] Event driven WiFi for ESP32

Not tested, just got some inspiration by these files:

- https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/examples/WiFiClientEvents/WiFiClientEvents.ino
- https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/src/WiFiGeneric.cpp
- https://github.com/espressif/arduino-esp32/blob/master/tools/sdk/include/esp32/esp_event.h

* automaticly updated release notes for mega-20180327

* Fix nesting level check

* [Build] Cleanup/split builds using defines

See PR #1034 and #1040 made by @soif

* [build] Disabled a few PlatformIO profiles for now.

There were way too many.
Just placed them in comments and Travis had already built them all to test first.

* automaticly updated release notes for mega-20180328

* [Arduino IDE] Revert changes in include paths

Were changed to be able to test the PlatformIO feature to compile only files that are actually used.
Apparently the Arduino IDE cannot find the files anymore because that IDE needs the parent directory to be named as the main project file, not "src"

* automaticly updated release notes for mega-20180329

GHolli committed:
Merge pull request #2 from letscontrolit/mega

Downstream updates

TD-er committed:
Build issue with Arduino IDE

See https://github.com/letscontrolit/ESPEasy/issues/1188#issuecomment-377020065

GHolli committed:
Update _P049_MHZ19.ino

Clarification of pin connections.

nunofgs committed:
Replace ultrasonic plugin with NewPing lib

TD-er committed:
[issue #1201] MQTT#Disconnected event and MQTT stability improvements

While looking for issues regarding DeepSleep + MQTT and issue #1195, I came across a number of other issues related to high CPU load and WiFi disconnects when a MQTT controller cannot reach the MQTT broker or is set to be no longer enabled.

These fixes improve CPU load and WiFi stability and also send the following events:
- `MQTT#Connected` 
- `MQTT#Disconnected`
- `MQTTimport#Connected` 
- `MQTTimport#Disconnected`

The MQTTimport plugin still has some major issues which have to be taken care of, so #1195 is not fixed yet.

TD-er committed:
[issue #1201] MQTT#Disconnected event and MQTT stability improvements

While looking for issues regarding DeepSleep + MQTT and issue #1195, I came across a number of other issues related to high CPU load and WiFi disconnects when a MQTT controller cannot reach the MQTT broker or is set to be no longer enabled.

These fixes improve CPU load and WiFi stability and also send the following events:
- `MQTT#Connected` 
- `MQTT#Disconnected`
- `MQTTimport#Connected` 
- `MQTTimport#Disconnected`

The MQTTimport plugin still has some major issues which have to be taken care of, so #1195 is not fixed yet.

psy0rz committed:
automaticly updated release notes for mega-20180330

s0170071 committed:
Serial Swap function

swaps Serial Pins *and* keeps track of it.

papperone commented:
thansk TD-er, so it will be a complete separate setup for the SONOFF POW but more pluings can be uses in many scenario:
- I've several PCBs with HLW8012 which are not POW and uses other devices/sensors, including a power plug with OLED display (using the framed version)
- On the market there's already one module with HLW8012 built by ElectroDragon that offers the same features and can be added to any ESP8266 configuration
what can be done to avoid this big limit of havin for POW switches/P075 only ?

TD-er commented:
It is not a hard limit, just an initial use-case.
Problem is the plugin uses (too) much iRAM, and thus may cause build failures when using with other plugins. So it cannot be used in generic "normal" or "dev" builds.
But for now, it can be used and tested and maybe we can later improve on it to use it in the generic builds.
So it is just a small D-tour to make progress instead of keep waiting.
And also an experiment for the manageability of separate builds for specific targets. Maybe it will be popular, maybe not.

I won't stop you in using it in other builds :)

martinusGH committed:
ESP32: Fix Wifi Disconnect reason

giig1967g committed:
Fixes for #1208 and partial #1196

Fixed #1208
Added dìformat commands:
#C: 1/0 = Open/Close
#U: 1/0 = Up/Down

TD-er committed:
[Build ESP32] Fix for build issues between ESP32 and ESP8266

There is a problem with the PlatformIO builder since the last update of @mvdbro
It's an issue with the parsing of `#ifdef ESP32`.

It is very well explained here: [platformio/platformio-core#801 (comment)](https://github.com/platformio/platformio-core/issues/801#issuecomment-252447178)

In short, the event handling functions for the wifi have been moved to a separate .h file, since those files will be parsed properly by the Library Dependency Finder (LDF) in PlatformIO 3.x

TD-er committed:
[ESP32] Build errors/warnings

TD-er committed:
Merge pull request #1212 from TD-er/bugfix/build_issues_ESP32_ESP8266

[Build ESP32] Fix for build issues between ESP32 and ESP8266

TD-er committed:
Merge remote-tracking branch 'origin/feature/mqtt_disconnect_event' into feature/mqtt_disconnect_event

psy0rz committed:
automaticly updated release notes for mega-20180331

papperone commented:
ok understood... but I'm still puzzled what is causign such huge usage of iRAM, did you had a chance to look at the library?

TD-er commented:
Not yet, but first some testing and stability issues.
I guess when this takes a lot of interrupts you may not want to have a lot of other plugins on the same node.

papperone commented:
but I don't use interrupts...
well I think if I've some time in the next days I'll try to incorporate the library removing the code related to interrupts as I'm not anyhow using it... this should solve the iRAM issue, correct?

TD-er commented:
I'm not sure whether interrupts are used (haven't looked at the code of the included .h file yet), but that's often the cause of high iRAM usage. The interrupt handler function then has too much variables active in the iRAM.

Grovkillen committed:
P0059 (rotary encoder) and P0063 (ttp229, keypad)

These two plugins are not in test but in normal.

TD-er committed:
[deepsleep] Allow for some minimum wifi connect time

In Deepsleep mode the wifi connect time is now minimum of 6 seconds.
Also the first loop after connect, the 'awake' timer will reset and start over. Without WiFi connection the set 'awake' timer starts at the moment the node boots.

TD-er committed:
Merge pull request #1203 from TD-er/feature/mqtt_disconnect_event

[issue #1201] MQTT#Disconnected event and MQTT stability improvements

TD-er committed:
Merge pull request #1205 from s0170071/SerialSwap

Serial Swap function

TD-er committed:
Revert "Merge pull request #1205 from s0170071/SerialSwap"

This reverts commit 46586e5355c9ed79190a252b8d09b0ce0dd74b48.

TD-er committed:
Merge remote-tracking branch 'origin/feature/p075_hlw8012' into feature/p075_hlw8012

psy0rz committed:
automaticly updated release notes for mega-20180401

psy0rz committed:
automaticly updated release notes for mega-20180402

soif committed:
Show CPU Load in JSON page

soif committed:
Adds CPU Load LC in JSON page

soif committed:
unwanted lib_ignore now properly commented

martinusGH committed:
ESP32 prep for RTOS multitasking

Grovkillen committed:
Update _P038_NeoPixel.ino

Grovkillen committed:
Update _P029_Output.ino

TD-er committed:
[issue #1215] Cannot login to accesspoint on new install

Since event based wifi, it became more apparent the ESP8266 has some issues with running AP and STA mode simultaneous.
It appears this can only work when the channel remains constant. But when trying to connect to an accesspoint that channel will change.
This means Soft AP and STA cannot be active at the same time when searching for an accesspoint to connect to. For example login to the Soft AP to correct some settings or replace an accesspoint would not be possible.

This fix deals with all of these issues and keeps the fast try/fail connection attempts offered by the event based wifi.
In short:
- 3x attempt to connect to AP1 (each attempt takes about 2.5 seconds when failing)
- 3x attempt to connect to AP2
- Enable accesspoint

When accesspoint is active, the connect attempts will be done at a slower pace and STA mode is disabled. STA mode is also disabled as soon as someone connects to the Soft AP.
When the settings are saved, a retry to connect like normal is started almost immediately. N.B. This will result in a disconnect of the connected users to the Soft AP.

So the connection time is:
- AP1 is valid: 4 - 10 seconds
- AP1 is not valid, AP2 is: 12 - 20 seconds
- AP1 and AP2 are not valid: +/- 18 seconds before activating Soft AP

TD-er committed:
[issue #1215] Cannot login to accesspoint on new install

Since event based wifi, it became more apparent the ESP8266 has some issues with running AP and STA mode simultaneous.
It appears this can only work when the channel remains constant. But when trying to connect to an accesspoint that channel will change.
This means Soft AP and STA cannot be active at the same time when searching for an accesspoint to connect to. For example login to the Soft AP to correct some settings or replace an accesspoint would not be possible.

This fix deals with all of these issues and keeps the fast try/fail connection attempts offered by the event based wifi.
In short:
- 3x attempt to connect to AP1 (each attempt takes about 2.5 seconds when failing)
- 3x attempt to connect to AP2
- Enable accesspoint

When accesspoint is active, the connect attempts will be done at a slower pace and STA mode is disabled. STA mode is also disabled as soon as someone connects to the Soft AP.
When the settings are saved, a retry to connect like normal is started almost immediately. N.B. This will result in a disconnect of the connected users to the Soft AP.

So the connection time is:
- AP1 is valid: 4 - 10 seconds
- AP1 is not valid, AP2 is: 12 - 20 seconds
- AP1 and AP2 are not valid: +/- 18 seconds before activating Soft AP

TD-er committed:
Merge remote-tracking branch 'origin/bugfix/SoftAP_event_based_wifi' into bugfix/SoftAP_event_based_wifi

psy0rz committed:
automaticly updated release notes for mega-20180403

TD-er committed:
[Wifi] Use statemachine to move between Wifi states

Still existing problems:
- Credentials are not being saved from web interface and thus connection doesn't work from setup page.
- First time connect to SoftAP takes a lot of attempts, when the wifi was searching for an existing accesspoint. This searching does prevent the SoftAP to form a stable connection to the client.

Maybe the entire wifi stack has to be reset before switching to SoftAP. Now it is near to impossible to recover from incorrect settings or a changes accesspoint.
Still under investigation.

It looks to be a problem with the

TD-er committed:
Merge pull request #1217 from soif/Feat/JsonLoad

Shows CPU Load in JSON page

TD-er committed:
Merge pull request #1218 from soif/Fix/PioConfTypo

Unwanted lib_ignore now properly commented

TD-er committed:
Merge pull request #1175 from Grovkillen/Fix-parsing-structure-for-JSON-+-upload-file-filter

Updated JSON path for sensor values and names

martinusGH committed:
ESP32 RTOS config setting

martinusGH committed:
ESP32 RTOS 10 per second

Run 10 per seconds call independent from main loop

giig1967g committed:
modified float() call

psy0rz committed:
automaticly updated release notes for mega-20180405

flexiti committed:
Merge pull request #1 from letscontrolit/mega

update

flexiti committed:
ESP CPU speed and ID

giig1967g committed:
new #formats

Hi, here the new formats:
#O: ON/OFF
#C: Open/Close
#U: UP/Down
#Y: YES/NO
#D: converto to nearest integer
#F: convert to floor
#E: convert to ceiling

giig1967g committed:
added more formats

#X
#y
#Z

Grovkillen committed:
Fixed clickable link www.letscontrolit.com (#1192)

giig1967g committed:
Added #I

#I: IN/OUT

GHolli committed:
Disable detection range commands by default.

psy0rz committed:
automaticly updated release notes for mega-20180407

TD-er committed:
[WiFi] Event based wifi, now AP mode works

It appeared to be related to the core 2.3.0 version. The 2.4.0 does just work. Also changed a number of other things to improve stability.

martinusGH committed:
Factory Reset (not enabled yet)

martinusGH committed:
Both reset/factoryreset option

TD-er committed:
[wifi] Make it work for ESP32 again

TD-er committed:
Merge branch 'mega' into bugfix/SoftAP_event_based_wifi

psy0rz committed:
automaticly updated release notes for mega-20180409

Grovkillen committed:
Added comments that will not show up on issues

Many users forget to "clean" the template from unnecessary text... these are now automatically hidden.

Grovkillen committed:
Update issue_template.md

Added a "Summarize" topic

Grovkillen committed:
Update issue_template.md

Added code format

Grovkillen committed:
Update issue_template.md

Rearranged the topics...

Grovkillen committed:
Update issue_template.md

`YOUR TEXT GOES HERE` added

GHolli committed:
Merge pull request #3 from letscontrolit/mega

Downstream updates from head fork.

psy0rz committed:
automaticly updated release notes for mega-20180410

Grovkillen committed:
Merge pull request #1 from letscontrolit/mega

Update

Commit:
add missing USES_XXX to _C*.ino, _N*.ino

Grovkillen committed:
Added RSSI, BSSID, Channel to WifiScan/Setup

And also delay is now renamed interval.

TD-er committed:
[wifi] Fix wifi AP setup mode for ESP32.

martinusGH committed:
Reboot/Reset pin, build 20101 patch

martinusGH committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

TD-er committed:
[MQTT] Process MQTT loop on publish

As [discussed here](https://github.com/letscontrolit/ESPEasy/issues/1228#issuecomment-380420609).

TD-er committed:
[wifi] Event based wifi, improve scan information.

TD-er committed:
Merge remote-tracking branch 'upstream/mega' into bugfix/SoftAP_event_based_wifi

TD-er committed:
Merge pull request #1220 from TD-er/bugfix/SoftAP_event_based_wifi

[issue #1215] Cannot login to accesspoint on new install

TD-er committed:
Merge pull request #1224 from flexiti/mega

ESP32 cpu speed and Id

psy0rz committed:
automaticly updated release notes for mega-20180412

Grovkillen committed:
Merge branch 'mega' into Fix-nomenclature-for-delay/interval

Grovkillen committed:
Removed the serial print "done."

Commit:
system variables %sysheap%, %syssec_d

- %sysheap%: free heap memory
- %syssec_d%: seconds since midnight

TD-er committed:
[wifi] Event based wifi, fix set AP and crash on start

The settings had 2 new values which would be overwriten by the checksum and there was still some issue about when to save the settings.

Also added a check to only save changed settings.

Known issue: 
Sometimes after initial setup of wifi credentials, it is best to perform a reset.

TD-er committed:
Merge pull request #1234 from TD-er/bugfix/Wifi_connection_issues

[wifi] Event based wifi, fix set AP and crash on start

psy0rz committed:
automaticly updated release notes for mega-20180413

Grovkillen committed:
Added better(?) format for tables

Fixed "normal" and "multirow" table css and made some column widths fixed.

Grovkillen committed:
Fixed main page!

Table was messed up... also update other tables

TD-er committed:
[issue #1239] %bssid% and %wi_ch% added as variables

See #1239

TD-er committed:
Merge pull request #1210 from giig1967g/mega

Fixes for #1208 and partial #1196

TD-er committed:
Merge pull request #1193 from clumsy-stefan/patch-1

_P015_TSL2561 added Broadband and IR Values

TD-er committed:
Merge pull request #1230 from Grovkillen/Fix-nomenclature-for-delay/interval

Renamed delay to interval/time for better understanding

TD-er committed:
Merge pull request #1240 from TD-er/feature/bssid_channel_variables

[issue #1239] %bssid% and %wi_ch% added as variables

TD-er committed:
Merge pull request #1078 from GHolli/mega

Sensor MH-Z19: Fixing offsets for the detection range setting command

TD-er committed:
Merge pull request #1186 from linuxnico/mega

add ds1822 definition to ds_read_temp function

psy0rz committed:
automaticly updated release notes for mega-20180414

Grovkillen committed:
Merge pull request #2 from letscontrolit/mega

Update source from MAIN

workgroupengineering committed:
Add Commands timerpause and timerresume

Commit:
ignore power-on reset value (85.0 °C)

TD-er committed:
Merge pull request #1243 from susisstrolch/defineUses

add missing USES_XXX to _C*.ino, _N*.ino

workgroupengineering committed:
Add Trigger Rules#TimerPause= and Rules#TimerResume=

TD-er committed:
Added C013

TD-er committed:
Merge pull request #1242 from susisstrolch/stringConverter

system variables %sysheap%, %syssec_d

TD-er committed:
Merge pull request #1209 from nunofgs/enhancement/replace-ultrasonic-plugin-with-newping-lib

Replace ultrasonic plugin with NewPing lib

Grovkillen committed:
Made checkmarks CSS

The &nbsp; is made to be able to have text after the custom checkbox... the text itself is not yet styled since we're not using labels like that yet.

Grovkillen committed:
Fixed missed update.

Grovkillen committed:
Added column width for separator...

Now variable

Commit:
DS_readTemp: handle power-on reset value

return false if DS_readTemp is power-on reset value
(0xaa for DS1820|DS18B20, 0x550 for DS18B20|DS1825|DS1822)

TD-er committed:
Merge pull request #1245 from susisstrolch/ds18b20

ignore power-on reset value (85.0 °C)

Grovkillen committed:
Updated some more stuff

Made the call for the function work without the extra (not used) "extra text"... it's still in the base function for future use.

Grovkillen committed:
Made addFormSeparator function a lot cleaner

Thanks to da man @TD-er

TD-er committed:
Merge pull request #1247 from Grovkillen/Fix-some-html

Fix some html

Grovkillen committed:
Bumped delay timer for MQTT pub 10mSec

Tests showed that instead of publishing "now" a 10mSec delay made the unit respond better and publish faster.

TD-er committed:
[WebServer] Cleanup of all functions appending to a big string

A lot of functions in Webserver.ino were just appending to a ever increasing string. These strings were eventually sent to TXBuffer of the Webinterface.
So why not directly send them to the streaming buffer and thus reduce memory usage and gain speed.

TD-er committed:
[PR #1247 ]  Merged PR into the refactoring of Webserver

TD-er committed:
[PR #1248] Fixed textarea sizes

See #1248

TD-er committed:
[Merge] Easier to merge, rebased on Mega branch.

TD-er committed:
Merge pull request #1249 from TD-er/bugfix/Webserver_appendingStrings

A lot of functions in Webserver.ino were just appending to a ever increasing string. These strings were eventually sent to TXBuffer of the Webinterface.
So why not directly send them to the streaming buffer and thus reduce memory usage and gain speed.

psy0rz committed:
automaticly updated release notes for mega-20180415

TD-er committed:
[wifi] Prevent starting AP+STA mode with correct settings

WiFi persistent settings were disabled after initialisation of the event handler.
This allowed the firmware to start WiFi connection with settings stored in the flash (not our settings, the settings kept by the core library). If these fail, then a connection failed event would occur before a connect attempt was started from ESPeasy, which would lead to strange results on some nodes while others work just fine.

TD-er committed:
[Wifi] force STA mode at boot

I finally have one node acting as described by others and this one reported a disconnect before a connect attempt was made.

Now forcing the wifi in STA mode, should resolve this issue.

TD-er committed:
[wifi] Proper log of static IP configuration

TD-er committed:
[info] Added FW information to help diagnose issues

A lot of issue reports do not include information about the used version of the core libraries and 'normal', 'dev', etc. 
Since this information helps narrowing down the possible causes of the reported issues, it is best to add them in the general log.

TD-er committed:
Merge pull request #1250 from TD-er/bugfix/wifi_start_ap_mode

[wifi] Prevent starting AP+STA mode with correct settings

psy0rz committed:
automaticly updated release notes for mega-20180416

Grovkillen committed:
Put borders around checkboxes

Suggested here: #1253

Grovkillen committed:
Re-arranged Rules page

Button moved down
Edit text removed

clumsy-stefan committed:
Changes for use of plugin sets

Remove ifdefs for PLUGIN_BUILD_TESTING to make use of USES_ defines and define_plugin_defs.h

Grovkillen committed:
Moved help button for sleep

Grovkillen committed:
Moved help button on hardware page

Moved it up (from right of the submit button which wasn't a good place for it)

Grovkillen committed:
Made controller symbols a bit bigger (1)(2)(3)

Grovkillen committed:
Change enable "off" to a nicer X

Grovkillen committed:
Made input type password part of "wide"

Grovkillen committed:
Added number input to CSS (width 100px)

Grovkillen committed:
Made IDX column wider (device list)

Grovkillen committed:
Added max widths for inputs

Grovkillen committed:
"Ctr (IDX)" was accidentally removed, now added

TD-er committed:
Merge pull request #1259 from Grovkillen/Updating-gui

Updating gui (max width etc)

TD-er committed:
[wifi] Fix wifi reconnects with static IP

When a fixed IP setting is used, there will be no "Got IP" event, which will lead to a timeout and thus reconnect.
As a fix, the "Connected" event will also act like when a "Got IP" event was received when configured to use static IP.

TD-er committed:
[NTP] Add more information in the log when connecting to NTP server fail

TD-er committed:
Merge pull request #1261 from TD-er/bugfix/fix_disconnects_static_ip

[wifi] Fix wifi reconnects with static IP

TD-er committed:
[issue #1260] parentheses error on the json page

TD-er committed:
Merge pull request #1262 from TD-er/bugfix/json_oled

[issue #1260] parentheses error on the json page

TD-er committed:
[P075] Nextion plugin (testing)

Added Nextion plugin for testing from the Playground.
It is not yet tested on real hardware, but it has been in development for a while on the forum and playground.
Also requested in issue #1256 and by several others.

TD-er committed:
[wifi] Fix build error for ESP32

Shared function was declared within #ifdef scope.

TD-er committed:
Merge pull request #1264 from TD-er/feature/nextion_port_from_playground

Feature/nextion port from playground

psy0rz committed:
automaticly updated release notes for mega-20180417

Grovkillen committed:
Merge pull request #1 from letscontrolit/mega

Updated from main

Grovkillen committed:
Updated the input, select, textareas with CSS

sentinel073 committed:
Issue #1211: Plugin template file and hints on writing new plugins

hansrune committed:
Implemented syslog facilities and levels

Sending LOG_LEVEL_ERROR as syslog error, LOG_LEVEL_INFO as syslog notice, and everything else as syslog deug
Also added Settings.Syslogfacility setting to be able to choose kernel (still default), user, daemon, message and local0-local7

Grovkillen committed:
Added CSS radio buttons to WiFi setup wizard

Grovkillen committed:
Removed unused row on hardware settings page

sentinel073 committed:
Issue #1211: Plugin template file and hints on writing new plugins
Added review comments

TD-er committed:
[Static IP] Work-around for set _use_static_ip

See [this bug report](https://github.com/esp8266/Arduino/issues/4114), which is not fixed in 2.4.0.

This work-around will make sure no dhcp-client will be started when connecting to a network.
The active DHCP client could result in IP conflicts, difficulty in reconnect and other strange issues.

TD-er committed:
Merge pull request #1275 from TD-er/bugfix/Static_ip

[Static IP] Work-around for set _use_static_ip

TD-er committed:
Merge pull request #1266 from Grovkillen/CSS-the-reset-of-the-gui

CSS the rest of the GUI

TD-er committed:
Merge pull request #1268 from hansrune/mega

Implemented syslog facilities and levels

TD-er committed:
Merge pull request #1267 from sentinel073/plugin_template_issue_1211

Issue #1211: Plugin template file and hints on writing new plugins

TD-er committed:
[PR #1257] Make it compile again

TD-er committed:
[ESP32] Add Generic info plugin

TD-er committed:
Merge pull request #1276 from TD-er/bugfix/pr_1257

Bugfix/pr 1257

psy0rz committed:
automaticly updated release notes for mega-20180419

Grovkillen committed:
Merge pull request #2 from letscontrolit/mega

update

Grovkillen committed:
Added small fade in (0.5s)

To make user understand that submit have been clicked (web interface is so fast now).

Grovkillen committed:
"bodymenu" was accidentally copied to all body

TD-er committed:
[Static IP] Make sure static mode is set correct always

Also when changing settings from static to DHCP and vice verse.

TD-er committed:
Merge pull request #1278 from TD-er/bugfix/set_static_at_change_mode

[Static IP] Make sure static mode is set correct always

Grovkillen committed:
Merge pull request #1277 from Grovkillen/Update-GUI

Update GUI a bit more

Grovkillen committed:
Added extra breaks below "back to setup" button

workgroupengineering committed:
Add support at elseif statment

TD-er committed:
[wifi] Restart webserver when connection changes.

TD-er committed:
[log] Make web log 30 lines for ESP32

TD-er committed:
[wifi] Create event 'gotIP' after processing 'connected' event

And stop webserver when trying to reconnect to WiFi

TD-er committed:
Merge pull request #1283 from TD-er/bugfix/webserver_restart_connection_change

Bugfix/webserver restart connection change

psy0rz committed:
automaticly updated release notes for mega-20180420

workgroupengineering committed:
Add debug info

Grovkillen committed:
Merge pull request #3 from letscontrolit/mega

Update

Grovkillen committed:
Removed fade in of body added toast message!

TD-er committed:
Merge pull request #1246 from workgroupengineering/feature/timer_pause_resume

[Feature] Timers Pause and Resume

TD-er committed:
Merge pull request #1285 from Grovkillen/toast-mes

Toast message added

TD-er committed:
Merge pull request #1287 from workgroupengineering/feature/elseif

[Feature] Allow elseif

TD-er committed:
[WiFi] Init static IP before connect to wifi

As suggested by @Feuerreiter  [here](https://github.com/letscontrolit/ESPEasy/issues/1255#issuecomment-383094333)

TD-er committed:
[Toast message] Change "Settings saved" to "Submitted"

Message is sent to provide feedback on pressing the 'submit' button.
At that moment, it is not yet clear the data will be saved successful.
To prevent "Settings Saved" and "Error writing to..." to show on the same page, this text was changed.

TD-er committed:
[info] Add more build related info to the log and wifi info link to wiki

Add core information to the log and info page
Add "more info" button to main page
Add question mark link button to sysinfo section describing wifi. This links to the wifi page on the wiki.

TD-er committed:
Merge pull request #1288 from TD-er/bugfix/InitStaticIP_before_connect

[WiFi] Init static IP before connect to wifi

TD-er committed:
[Sonoff POW] Change Plugin ID to 76

Nextion was added as P075

TD-er committed:
Merge remote-tracking branch 'origin/feature/p075_hlw8012' into feature/p075_hlw8012

TD-er commented:
Superseded by #1290
